File: typescript/src/dtype.ts
/** The dtype table: the element types a tensor can hold, and their widths in bytes. */
export const DTYPE_WIDTHS = {
  F64: 8,
  F32: 4,
  F16: 2,
  BF16: 2,
  I64: 8,
  I32: 4,
  I16: 2,
  I8: 1,
  U64: 8,
  U32: 4,
  U16: 2,
  U8: 1,
  BOOL: 1,
} as const;

export type Dtype = keyof typeof DTYPE_WIDTHS;

/** Every dtype, in the order the table lists them. */
export const DTYPES = Object.keys(DTYPE_WIDTHS) as readonly Dtype[];

/** Whether a string names a dtype. */
export function isDtype(name: string): name is Dtype {
  return Object.hasOwn(DTYPE_WIDTHS, name);
}

/**
 * The typed array a dtype decodes to. `BF16` has no JavaScript type, so it
 * decodes as the `Uint16Array` of its bit patterns, and `BOOL` as a
 * `Uint8Array` of zeros and ones.
 */
export type TypedArrayOf<D extends Dtype> = D extends "F64"
  ? Float64Array
  : D extends "F32"
    ? Float32Array
    : D extends "F16"
      ? Float16Array
      : D extends "BF16"
        ? Uint16Array
        : D extends "I64"
          ? BigInt64Array
          : D extends "I32"
            ? Int32Array
            : D extends "I16"
              ? Int16Array
              : D extends "I8"
                ? Int8Array
                : D extends "U64"
                  ? BigUint64Array
                  : D extends "U32"
                    ? Uint32Array
                    : D extends "U16"
                      ? Uint16Array
                      : D extends "U8"
                        ? Uint8Array
                        : D extends "BOOL"
                          ? Uint8Array
                          : never;

export type TypedArray = TypedArrayOf<Dtype>;

interface TypedArrayConstructor {
  readonly BYTES_PER_ELEMENT: number;
  new (buffer: ArrayBufferLike, byteOffset?: number, length?: number): TypedArray;
}

export const TYPED_ARRAYS: Record<Dtype, TypedArrayConstructor> = {
  F64: Float64Array,
  F32: Float32Array,
  F16: Float16Array,
  BF16: Uint16Array,
  I64: BigInt64Array,
  I32: Int32Array,
  I16: Int16Array,
  I8: Int8Array,
  U64: BigUint64Array,
  U32: Uint32Array,
  U16: Uint16Array,
  U8: Uint8Array,
  BOOL: Uint8Array,
};

/** The dtype a typed array naturally holds, with `Uint16Array` taken as `U16` and `Uint8Array` as `U8`. */
export function dtypeOf(array: TypedArray): Dtype {
  if (array instanceof Float64Array) return "F64";
  if (array instanceof Float32Array) return "F32";
  if (array instanceof Float16Array) return "F16";
  if (array instanceof BigInt64Array) return "I64";
  if (array instanceof Int32Array) return "I32";
  if (array instanceof Int16Array) return "I16";
  if (array instanceof Int8Array) return "I8";
  if (array instanceof BigUint64Array) return "U64";
  if (array instanceof Uint32Array) return "U32";
  if (array instanceof Uint16Array) return "U16";
  return "U8";
}

/** The `bfloat16` bit patterns of a `Uint16Array` as single-precision floats. */
export function bf16ToFloat32(bits: Uint16Array): Float32Array {
  const out = new Float32Array(bits.length);
  const view = new DataView(out.buffer);
  for (let i = 0; i < bits.length; i++) {
    view.setUint32(i * 4, bits[i]! << 16, true);
  }
  return out;
}

/** Single-precision floats as `bfloat16` bit patterns, rounded to nearest even. */
export function float32ToBf16(values: Float32Array): Uint16Array {
  const out = new Uint16Array(values.length);
  const bits = new Uint32Array(values.buffer, values.byteOffset, values.length);
  for (let i = 0; i < values.length; i++) {
    const word = bits[i]!;
    if ((word & 0x7fffffff) > 0x7f800000) {
      out[i] = (word >>> 16) | 0x40;
      continue;
    }
    const lower = word & 0xffff;
    const upper = word >>> 16;
    const roundUp = lower > 0x8000 || (lower === 0x8000 && (upper & 1) === 1);
    out[i] = (upper + (roundUp ? 1 : 0)) & 0xffff;
  }
  return out;
}

File: rust/src/dtype.rs
//! The dtype table: the element types a tensor can hold.

use std::fmt;

/// An element type. Every multi-byte value is little-endian in the buffer.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum Dtype {
    /// IEEE 754 binary64.
    F64,
    /// IEEE 754 binary32.
    F32,
    /// IEEE 754 binary16.
    F16,
    /// bfloat16, the high half of a binary32.
    BF16,
    I64,
    I32,
    I16,
    I8,
    U64,
    U32,
    U16,
    U8,
    /// One byte per element, 0 or 1.
    Bool,
}

impl Dtype {
    /// Every dtype, in the order the table lists them.
    pub const ALL: [Dtype; 13] = [
        Dtype::F64,
        Dtype::F32,
        Dtype::F16,
        Dtype::BF16,
        Dtype::I64,
        Dtype::I32,
        Dtype::I16,
        Dtype::I8,
        Dtype::U64,
        Dtype::U32,
        Dtype::U16,
        Dtype::U8,
        Dtype::Bool,
    ];

    /// The name a reference's `$dtype` carries.
    pub fn name(self) -> &'static str {
        match self {
            Dtype::F64 => "F64",
            Dtype::F32 => "F32",
            Dtype::F16 => "F16",
            Dtype::BF16 => "BF16",
            Dtype::I64 => "I64",
            Dtype::I32 => "I32",
            Dtype::I16 => "I16",
            Dtype::I8 => "I8",
            Dtype::U64 => "U64",
            Dtype::U32 => "U32",
            Dtype::U16 => "U16",
            Dtype::U8 => "U8",
            Dtype::Bool => "BOOL",
        }
    }

    /// The dtype a `$dtype` value names, if it names one.
    pub fn from_name(name: &str) -> Option<Dtype> {
        Dtype::ALL.into_iter().find(|dtype| dtype.name() == name)
    }

    /// The width of one element in bytes.
    pub fn width(self) -> u64 {
        match self {
            Dtype::F64 | Dtype::I64 | Dtype::U64 => 8,
            Dtype::F32 | Dtype::I32 | Dtype::U32 => 4,
            Dtype::F16 | Dtype::BF16 | Dtype::I16 | Dtype::U16 => 2,
            Dtype::I8 | Dtype::U8 | Dtype::Bool => 1,
        }
    }
}

impl fmt::Display for Dtype {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.name())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn names_round_trip() {
        for dtype in Dtype::ALL {
            assert_eq!(Dtype::from_name(dtype.name()), Some(dtype));
        }
        assert_eq!(Dtype::from_name("F128"), None);
        assert_eq!(Dtype::from_name("bool"), None);
    }

    #[test]
    fn widths_match_the_table() {
        assert_eq!(Dtype::F64.width(), 8);
        assert_eq!(Dtype::BF16.width(), 2);
        assert_eq!(Dtype::Bool.width(), 1);
    }
}
